Change `maybe_resize_storage_cpu` `new_size` arg to unsigned

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52672 `maybe_resize_storage_cuda` new_size argument should be unsigned
* **#52671 Change `maybe_resize_storage_cpu` `new_size` arg to unsigned**
* #52670 Fix wrong TORCH_CHECK usages

Code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.

Unroll nested if blocks by returning early if `new_size` is 0

Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

Differential Revision: [D26607549](https://our.internmc.facebook.com/intern/diff/D26607549)

malfet committed:
`maybe_resize_storage_cpu` to error on negative numbers

Current code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.
Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

[ghstack-poisoned]

malfet committed:
Update on "`maybe_resize_storage_cpu` to error on negative numbers"


Current code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.
Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

Differential Revision: [D26607549](https://our.internmc.facebook.com/intern/diff/D26607549)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Resize.h:
```diff
@@ -22,23 +22,24 @@ TORCH_API void resize_output(Tensor& output, IntArrayRef shape);
 // to benchmark than TH; I can't get gbenchmark to call fns from THTensor.cpp
 
-static inline void maybe_resize_storage_cpu(TensorImpl* self, int64_t new_size) {
+static inline void maybe_resize_storage_cpu(TensorImpl* self, uint64_t new_size) {
```
It makes sense that this function would take an unsigned integer, but this update doesn't seem to reflect the name of the PR.

This function's also exclusively called by resize_impl_cpu_, isn't the issue that it computes `storage_size` in int64_t and doesn't check for the size being representable in int64_t? Naively, I guess I would have expected an explicit check for new_size being less than zero or zero. 

Also, in the case where new_size == 0, what happens if the tensor doesn't have a storage pointer?

malfet replied:
Will update PR title (and also skip the test under asan, which correctly detects integer overflow).

And there is already a check for `new_size == 0`, which exists, according to https://github.com/pytorch/pytorch/commit/c46c6a4fe68a8ed6a8adb35837e87209cf020547  handle zero sized tensor slicing, i.e.:
```
python -c "import numpy as np;import torch;t=torch.from_numpy(np.empty((0,4)));t[:,1::2]*=1"
```

mruberry replied:
I know there's a check for new_size == 0, but if the storage is uninitialized it just stays uninitialized in that case, right?

malfet replied:
Yes, as far as I understand, this is the intended behaviour.

mruberry replied:
OK.

malfet committed:
Update on "`maybe_resize_storage_cpu` to error on negative numbers"


Current code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.
Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

Differential Revision: [D26607549](https://our.internmc.facebook.com/intern/diff/D26607549)

[ghstack-poisoned]

malfet committed:
Update on "Change `maybe_resize_storage_cpu` `new_size` arg to unsigned"


Code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.

Unroll nested if blocks by returning early if `new_size` is 0

Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

Differential Revision: [D26607549](https://our.internmc.facebook.com/intern/diff/D26607549)

[ghstack-poisoned]
